### PR TITLE
Vishal/increasing msg size limits

### PR DESCRIPTION
**Pull request by vishalchangrani**
Three changes-
1. Increase the pusub message size to 5mb from 2mb
2. Allow the size of the unicast message type `messages.ChunkDataResponse` to be up to 5GB
3. Changed `p2p.NewMiddleware` to no longer accept the pubsub and unicast message sizes as parameters but to just default to the constants defined in `middleware.go`

### CODE REVIEW

File: network/p2p/middleware.go [hunk 1]
```diff
@@ -32,17 +32,30 @@ const (
...
 
-	// maximum time to wait for a unicast request to complete
+	// default maximum time to wait for a default unicast request to complete
+	// assuming at least a 1mb/sec connection
 	DefaultUnicastTimeout = 2 * time.Second
-)
 
-var unicastTimeout = DefaultUnicastTimeout
+	// maximum time to wait for a unicast request to complete for large message size
+	LargeMsgUnicastTimeout = 1000 * time.Second
```
Comment (vishalchangrani) [inline network/p2p/middleware.go:57]: the `1gb` large message will take more time to transfer and hence a bigger unicast timeout. The time out is based on the previous timeout of 2 second for a 2 mb message (1mb/sec).

File: network/p2p/readConnection.go [hunk 2]
```diff
@@ -67,31 +67,54 @@ func (rc *readConnection) receiveLoop(wg *sync.WaitGroup) {
...
-				if err != nil {
-					rc.log.Error().Err(err)
-				}
+				rc.closeStream()
 			}
 			return
 		}
 
+		// check message size
+		maxSize := unicastMaxMsgSize(&msg)
```
Comment (vishalchangrani) [inline network/p2p/readConnection.go:86]: receive side check to reject large messages

File: network/p2p/middleware.go [hunk 3]
```diff
@@ -332,7 +333,7 @@ func (m *Middleware) handleIncomingStream(s libp2pnetwork.Stream) {
 	log.Info().Msg("incoming connection established")
 
 	//create a new readConnection with the context of the middleware
-	conn := newReadConnection(m.ctx, s, m.processMessage, log, m.metrics, m.maxUnicastMsgSize)
+	conn := newReadConnection(m.ctx, s, m.processMessage, log, m.metrics, LargeMsgMaxUnicastMsgSize)
```
Comment (vishalchangrani) [inline network/p2p/middleware.go:336]: The `readConnection` has to expect a message of upto `LargeMsgMaxUnicastMsgSize`. Once the message is received, its type can be used to determine if it exceeds size but that can only happen after it has been _fully_ received.
So, this does unfortunately introduce the possibility that a bad sender can keep sending large messages (other than `messages.ChunkDataResponse`) of size > 2mb and unnecessarily keep the received busy doing i/o.

File: network/p2p/middleware.go [hunk 4]
```diff
@@ -99,8 +100,6 @@ func NewMiddleware(log zerolog.Logger,
 		me:                flowID,
 		libP2PNodeFactory: libP2PNodeFactory,
 		metrics:           metrics,
-		maxPubSubMsgSize:  maxPubSubMsgSize,
```
Comment (vishalchangrani) [inline network/p2p/middleware.go:102]: just using the default constants instead

File: network/p2p/middleware.go (hunk 3 above)
Comment (yhassanzadeh13) [reply network/p2p/middleware.go:336]: Thanks for pointing this, I think it is worthy of opening an issue as part of _benign byzantine_ vulnerabilities of network layer.

File: network/p2p/middleware.go [hunk 5]
```diff
@@ -459,3 +460,23 @@ func (m *Middleware) UpdateAllowList() error {
...
+
+// unicastMaxMsgSize returns the max permissible size for a unicast message
+func unicastMaxMsgSize(msg *message.Message) int {
+	switch msg.Type {
+	case "messages.ChunkDataResponse":
+		return LargeMsgMaxUnicastMsgSize
+	default:
+		return DefaultMaxUnicastMsgSize
+	}
+}
```
Comment (yhassanzadeh13) [inline network/p2p/middleware.go:472]: 👍

File: network/p2p/readConnection.go [hunk 6]
```diff
@@ -67,31 +67,54 @@ func (rc *readConnection) receiveLoop(wg *sync.WaitGroup) {
...
+		if msg.Size() > maxSize {
+			// if message size exceeded, close stream and log error
+			rc.closeStream()
+			rc.log.Error().
```
Comment (yhassanzadeh13) [inline network/p2p/readConnection.go:90]: I would suggest escalating its level to `fatal`. If such a situation happens we'd better catch it in the snap of a finger than delving into the `error` level logs.

File: network/p2p/readConnection.go [hunk 7]
```diff
@@ -67,31 +67,54 @@ func (rc *readConnection) receiveLoop(wg *sync.WaitGroup) {
...
+				Hex("sender", msg.OriginID).
+				Hex("event_id", msg.EventID).
+				Str("event_type", msg.Type).
+				Str("channel_id", msg.ChannelID).
```
Comment (yhassanzadeh13) [inline network/p2p/readConnection.go:94]:
```suggestion
				Str("channel", msg.ChannelID).
```

File: network/p2p/readConnection.go [hunk 8]
```diff
@@ -67,31 +67,54 @@ func (rc *readConnection) receiveLoop(wg *sync.WaitGroup) {
...
+	err := rc.stream.Close()
+	if err != nil {
+		rc.log.Error().Err(err)
+	}
+}
+
+func (rc *readConnection) resetStream() {
+	err := rc.stream.Reset()
+	if err != nil {
+		rc.log.Error().Err(err)
```
Comment (yhassanzadeh13) [inline network/p2p/readConnection.go:118]: Please add a message to the logs on these two methods (e.g., `could not close stream`). It makes it drastically easier to spot the problem.

File: network/test/middleware_test.go [hunk 9]
```diff
@@ -261,6 +261,35 @@ func (m *MiddlewareTestSuite) TestMaxMessageSize_SendDirect() {
 	require.Error(m.Suite.T(), err)
 }
 
+// TestLargeMessageSize_SendDirect asserts that a large message can be unicasted successfully
```
Comment (yhassanzadeh13) [inline network/test/middleware_test.go:264]: I think the godoc is better to reflect more on the content of the test, we evaluate that a specific type of message that is our desire to be sent successfully although its size is beyond the limit.

**Review by yhassanzadeh13 — APPROVED** [review]
Thanks for the neat implementation, just some minor comments.

File: network/p2p/readConnection.go (hunk 6 above)
Comment (vishalchangrani) [reply network/p2p/readConnection.go:90]: i believe `level.fatal` tends to also throw a fatal error and bring down the node, doesn't it?

File: network/p2p/middleware.go (hunk 3 above)
Comment (vishalchangrani) [reply network/p2p/middleware.go:336]: https://github.com/dapperlabs/flow-go/issues/5191

File: network/p2p/readConnection.go [hunk 10]
```diff
@@ -67,24 +69,33 @@ func (rc *readConnection) receiveLoop(wg *sync.WaitGroup) {
...
-				if err != nil {
-					rc.log.Error().Err(err)
-				}
-			} else {
-				err = rc.stream.Close()
-				if err != nil {
-					rc.log.Error().Err(err)
-				}
+			// if the sender closes the connection, an EOF is received
+			if err == io.EOF {
```
Comment (vishalchangrani) [inline network/p2p/readConnection.go:78]: simplifying the indentation a little by rearranging the `if`

**Review by AlexHentschel — APPROVED** [review]
Looks great. Thanks Vishal for jumping on this on such a quick notice.

**Review by ramtinms — APPROVED** [review]
❤️ for the chunk data pack size

File: network/p2p/readConnection.go (hunk 6 above)
Comment (vishalchangrani) [reply network/p2p/readConnection.go:92]: I experimented with a `log.fatal` somewhere else in the code and it does stops the node. Hence, for now I am keeping it as an `error`.

**Pull request merged**